clientconn_test.go

```diff
@@ -820,3 +820,129 @@ func TestBackoffCancel(t *testing.T) {
+	if err != nil {
+		t.Fatalf("Error while listening. Err: %v", err)
+	}
+	defer lis.Close()
+	testFinished := make(chan struct{})
+	backoffCaseReady := make(chan struct{}, 1)
+	defer close(backoffCaseReady)
+	killFirstConnection := make(chan struct{})
+	killSecondConnection := make(chan struct{})
+
+	// Launch the server.
+	go func() {
+		// Establish a successful connection so that we enter READY. We need to get
+		// to READY so that we can get a client back for us to introspect later (as
+		// opposed to just CONNECTING).
+		conn, err := lis.Accept()
+		if err != nil {
+			t.Error(err)
+			return
+		}
+		defer conn.Close()
+
+		framer := http2.NewFramer(conn, conn)
+		if err := framer.WriteSettings(http2.Setting{}); err != nil {
+			t.Errorf("Error while writing settings frame. %v", err)
+			return
+		}
+
+		select {
+		case <-testFinished:
+			return
+		case <-killFirstConnection:
+		}
+
+		// Close the conn to cause onShutdown, causing us to enter TRANSIENT FAILURE. Note that we are not in
+		// WaitForHandshake at this point because the preface was sent successfully.
+		conn.Close()
+
+		// We have to re-accept and re-close the connection because the first re-connect after a successful handshake
+		// has no backoff. So, we need to get to the second re-connect after the successful handshake for our infinite
+		// backoff to happen.
+		conn, err = lis.Accept()
+		if err != nil {
+			t.Error(err)
+			return
+		}
+		err = conn.Close()
+		if err != nil {
+			t.Error(err)
+		}
+
+		backoffCaseReady <- struct{}{}
+
+		// Re-connect (without lis preface).
+		conn, err = lis.Accept()
+		if err != nil {
+			t.Error(err)
+			return
+		}
+
+		// Close the conn to cause onShutdown, causing us to enter TRANSIENT FAILURE. Note that we are in
+		// WaitForHandshake at this point because the preface has not been sent yet.
+		select {
+		case <-testFinished:
+			return
+		case <-killSecondConnection:
+		}
+		conn.Close()
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	client, err := DialContext(ctx, lis.Addr().String(), WithInsecure(), WithWaitForHandshake(), WithBlock(), withBackoff(backoffForever{}))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer client.Close()
+
+	// It should start in READY because the lis sends the lis preface.
+	if client.GetState() != connectivity.Ready {
+		t.Fatalf("expected addrconn to be in READY, was %v", client.GetState())
+	}
+
+	// Once the connection is killed, it should go:
+	// READY -> CONNECTING -> TRANSIENT FAILURE (no backoff) -> CONNECTING -> TRANSIENT FAILURE (infinite backoff)
+	// The first TRANSIENT FAILURE is triggered by closing a channel. Then, we wait for the lis to let us know
+	// when the client has progressed past the first failure (which does not get backoff, because handshake was
+	// successful).
+	close(killFirstConnection)
+	<-backoffCaseReady
+	client.WaitForStateChange(ctx, connectivity.TransientFailure)
```

Comment:
I don't think this line is necessary / may be racy.

We _should_ be in `Connecting` - or already in `TransientFailure` caused by the server closing the connection.  If we hit the latter case, this will block until `ctx` expires.

Reply:
Done.